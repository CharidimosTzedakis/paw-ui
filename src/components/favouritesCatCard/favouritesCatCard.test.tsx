import { describe, it, Mock, vi, expect, beforeEach } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import theCatAPI from "@api/catApiClient";
import FavouritesCatCard from "./favouritesCatCard";
import { useLocation } from "wouter";

vi.mock("wouter", async () => {
  const actual = await vi.importActual("wouter");

  return {
    ...actual,
    useLocation: vi.fn(() => ["/current-path", vi.fn()]),
  };
});

vi.mock("@api/catApiClient", () => ({
  default: {
    favourites: {
      deleteFavourite: vi.fn(),
    },
  },
}));

const FAVOURITES_ENTRY_ID = 123;

describe("<FavouritesCatCard />", () => {
  beforeEach(() => {
    vi.clearAllMocks(); // or vi.resetAllMocks()
  });

  it("should render correctly", () => {
    const { container } = render(
      <FavouritesCatCard
        favouritesId={FAVOURITES_ENTRY_ID}
        imageId="imageId"
        imageUrl="https://example.com/image"
        onFavouritesRemove={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should navigate to cat image details when "View details" button is clicked', async () => {
    const setLocation = vi.fn();
    (useLocation as Mock).mockReturnValue(["/current-path", setLocation]);

    render(
      <FavouritesCatCard
        favouritesId={FAVOURITES_ENTRY_ID}
        imageId="imageId"
        imageUrl="https://example.com/image"
        onFavouritesRemove={() => {}}
      />,
    );

    const closeButton = screen.getByText(/View details/i);
    await userEvent.click(closeButton);

    expect(setLocation).toHaveBeenCalledTimes(1);
    expect(setLocation).toHaveBeenCalledWith("../cats/imageId");
  });

  it('should remove the image from favourites when "Remove" button is clicked', async () => {
    const onFavouritesRemove = vi.fn();
    (theCatAPI.favourites.deleteFavourite as Mock).mockImplementation(() =>
      Promise.resolve({}),
    );

    render(
      <FavouritesCatCard
        favouritesId={FAVOURITES_ENTRY_ID}
        imageId="imageId"
        imageUrl="https://example.com/image"
        onFavouritesRemove={onFavouritesRemove}
      />,
    );

    const removeButton = screen.getByText(/Remove/i);
    await userEvent.click(removeButton);

    await waitFor(() =>
      expect(
        screen.getByText(
          "Are you sure you want to remove this image from your favourites?",
        ),
      ).toBeInTheDocument(),
    );

    const confirmButton = screen.getByRole("button", { name: /yes/i });
    await userEvent.click(confirmButton);

    await waitFor(() => expect(onFavouritesRemove).toHaveBeenCalledTimes(1));
    expect(theCatAPI.favourites.deleteFavourite).toHaveBeenCalledTimes(1);
    expect(theCatAPI.favourites.deleteFavourite).toHaveBeenCalledWith(
      FAVOURITES_ENTRY_ID,
    );
  });

  it('should not perform removal of favourite image if "No" is pressed during confirmation', async () => {
    const onFavouritesRemove = vi.fn();
    (theCatAPI.favourites.deleteFavourite as Mock).mockImplementation(() =>
      Promise.resolve({}),
    );

    render(
      <FavouritesCatCard
        favouritesId={FAVOURITES_ENTRY_ID}
        imageId="imageId"
        imageUrl="https://example.com/image"
        onFavouritesRemove={onFavouritesRemove}
      />,
    );

    const removeButton = screen.getByText(/Remove/i);
    await userEvent.click(removeButton);

    await waitFor(() =>
      expect(
        screen.getByText(
          "Are you sure you want to remove this image from your favourites?",
        ),
      ).toBeInTheDocument(),
    );

    const cancelButton = screen.getByRole("button", { name: /no/i });
    await userEvent.click(cancelButton);

    await waitFor(() => expect(onFavouritesRemove).not.toHaveBeenCalled());
    expect(theCatAPI.favourites.deleteFavourite).not.toHaveBeenCalled();
  });
});
