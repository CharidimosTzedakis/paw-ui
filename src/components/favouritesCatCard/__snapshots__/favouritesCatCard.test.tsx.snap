// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<FavouritesCatCard /> > should render correctly 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered _favouritesCatCard_ca1681 css-dev-only-do-not-override-1kf000u"
  >
    <div
      class="ant-card-body"
    >
      <div
        class="ant-image css-dev-only-do-not-override-1kf000u"
      >
        <img
          class="ant-image-img _favouriteImage_ca1681 css-dev-only-do-not-override-1kf000u"
          src="https://example.com/image"
        />
        <div
          class="ant-image-mask"
        >
          <div
            class="ant-image-mask-info"
          >
            <span
              aria-label="eye"
              class="anticon anticon-eye"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="eye"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                />
              </svg>
            </span>
            Preview
          </div>
        </div>
      </div>
      <div
        class="_actions_ca1681"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-1kf000u ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
          tabindex="0"
          type="button"
        >
          <span>
            View details
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-1kf000u ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
          tabindex="0"
          type="button"
        >
          <span
            aria-label="delete"
            class="anticon anticon-delete"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="delete"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
              />
            </svg>
          </span>
          <span>
            Remove
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
